src/frontends/common/include/openvino/frontend/shared_object_extension.hpp

```diff
@@ -0,0 +1,129 @@
+// Copyright (C) 2018-2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+#pragma once
+
+#include <algorithm>
+#include <memory>
+
+namespace ov {
+namespace frontend {
+
+/**
+ * \brief Extension to Derived class by adding shared object smart pointer.
+ *
+ * Class provide helper methods to make smart pointers from  pointer return by pimpl and shared object which can
+ * provide correct life time of required shared object (library).
+ *
+ * Rationale:
+ *
+ * To correct use the pimpl pattern with dynamically loaded libraries the class which has got the pimpl has to have
+ * pointer to library which is required by pimpl. All methods which use return values (smart pointers) from pimpl
+ * cannot return them directly. These pointers and shared object has to be used as input data for new smart pointers
+ * for type which has got pimpl member.
+ *
+ * Smart pointer returned directly from pimpl can ensure enough life time for library to destroy object pointed by
+ * smart pointer but will fail on destroy of smart pointer object.
+ *
+ * \tparam Derived  Class which get features from SharedObjectExtension.
+ */
+template <class Derived>
+class SharedObjectExtension {
```

Comment:
Can we remove this object from public API?

Reply:
Is not by default hidden unless `FRONTEND_API` or similar attribute will be added?